File: structy/02-linked-list/04-get-node-value/02.js
/*
get node value
Write a function, getNodeValue, that takes in the head of a linked list and an index. The function should return the value of the linked list at the specified index.

If there is no node at the given index, then return null.
*/

/*
INPUT: Node 
OUTPUT: string | null

RULES 
- it should return the value of the linked list at the specified index 
- if there's no node at the given index, then return null

ALGORITHM 
- Given the function `getNodeValue(head: Node, index: number)`

- if `index` is less than 0 
    - return NULL

- set `currentIndex` to 0
- set `currentNode` to `head`
- while `currentNode` is not NULL 
    - if `currentIndex` === `index`
        - return `currentNode`'s `val` property value 
    - increment `currentIndex` by 1 
    - set `currentNode` to currentNode.next
- return NULL 

COMPLEXITY 
- N: the number of nodes in the linked list 
- Time: O(N)
- Space: O(1)
*/
import Node from '../node.js';

const getNodeValue = (head, targetIndex, currentIndex=0) => {
    if (head === null) {
        return null
    }

    if (targetIndex === currentIndex) {
        return head.val;
    }

    return getNodeValue(head.next, targetIndex, currentIndex + 1);
}


const a = new Node("a");
const b = new Node("b");
const c = new Node("c");
const d = new Node("d");

a.next = b;
b.next = c;
c.next = d;

// a -> b -> c -> d
console.log(getNodeValue(a, 0)); // null

File: structy/02-linked-list/04-get-node-value/02.py
# get node value
# Write a function, get_node_value, that takes in the head of a linked list and an index. The function should return the value of the linked list at the specified index.

# If there is no node at the given index, then return None.

"""
INPUT:
- head: Node
- index: int

OUTPUT: 
- Any (The value at the specified index)

RULES 
- it should return the value of the linked list at the specified index
- if the input index is larger than the number of nodes in the linked list, return None
- Assumptions
    - assume only positive integers will be provied as input


EXAMPLES
T1
input: "red" -> "green" -> "blue" -> None, 2
output: "blue"

T2
input: 1 -> 2 -> 3 -> 4 -> None, 1
output: 2

T3
input: head=1 -> 2 -> 3, index=29
output: None

T4
input: head=1 -> 2 -> 3 -> 4 -> None, index=9
output: None

MENTAL MODEL 
- Given a linked list and an index position, return the value of the node at that index position.

SCENARIOS
Given a valid linked list:

1. positive index within the appropriate range is provided 
    - return the value of the node at that index position
2. positive index outside the range is provided 
    - return None
3. The input index is 0 
    - return the value at index 0
4 The input is other than 0
    - 

ALGORITHM
Iterative Algorithm
- set index to 0 
- while current node is not NULL
    - if index is equal to input index 
        - return node.val
    - increment index by 1
    - advance the current node
- return None

COMPLEXITY
Iterative Solution
- N: length of LL
- Time: O(N)
- Space: O(1)

Recursive Algorithm 
given: 
- head: Node
- index: int

- if head is None:
    - return None
- if index is less than or equal to 0
    - return head.val
- return recursive call with `head.next` and `index - 1`


Recursive Solution 
- N: length of LL
- Time: O(N)
- Space: O(N)
"""

class Node:
    def __init__(self, val):
        self.val = val
        self.next = None


# def get_node_value(head: Node, index: int):
#     current_index = 0
#     current_node = head 
#     while current_node is not None:
#         if current_index == index:
#             return current_node.val
#         current_node = current_node.next
#         current_index += 1
#     return None



def get_node_value(head: Node, index: int):
    if head is None:
        return None
    if index <= 0:
        return head.val
    return get_node_value(head.next, index - 1)


a = Node("a")
b = Node("b")
c = Node("c")
d = Node("d")

a.next = b
b.next = c
c.next = d

# a -> b -> c -> d

print(get_node_value(a, 2)) # 'c'


a = Node("a")
b = Node("b")
c = Node("c")
d = Node("d")

a.next = b
b.next = c
c.next = d

# a -> b -> c -> d

print(get_node_value(a, 7)) # None


node1 = Node("banana")
node2 = Node("mango")

node1.next = node2

# banana -> mango

print(get_node_value(node1, 1)) # 'mango'